modules/etcd/etcd.go

```diff
@@ -0,0 +1,285 @@
+	// and/or check that there are no child nodes
+	if c.opts.clusterNetwork != nil && c.opts.currentNode == 0 {
+		if err := c.opts.clusterNetwork.Remove(ctx); err != nil {
+			errs = append(errs, fmt.Errorf("remove cluster network: %w", err))
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+// Run creates an instance of the etcd container type
+func Run(ctx context.Context, img string, opts ...testcontainers.ContainerCustomizer) (*EtcdContainer, error) {
+	req := testcontainers.ContainerRequest{
+		Image: img,
+		Cmd:   []string{},
+	}
+
+	genericContainerReq := testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	}
+
+	settings := defaultOptions(&req)
+	for _, opt := range opts {
+		if apply, ok := opt.(Option); ok {
+			apply(&settings)
+		}
+		if err := opt.Customize(&genericContainerReq); err != nil {
+			return nil, err
+		}
+	}
+
+	clusterOpts, err := configureCluster(ctx, &settings, opts)
+	if err != nil {
+		return nil, fmt.Errorf("configure cluster: %w", err)
+	}
+
+	// configure CMD with the nodes
+	genericContainerReq.Cmd = configureCMD(settings)
+
+	// Initialise the etcd container with the current settings.
+	// The cluster network, if needed, is already part of the settings,
+	// so the following error handling returns a partially initialised container,
+	// allowing the caller to clean up the resources with the Terminate method.
+	c := &EtcdContainer{opts: settings}
+
+	if settings.clusterNetwork != nil {
+		// apply the network to the current node
+		err := tcnetwork.WithNetwork([]string{settings.nodeNames[settings.currentNode]}, settings.clusterNetwork)(&genericContainerReq)
+		if err != nil {
+			return c, fmt.Errorf("with network: %w", err)
+		}
+	}
+
+	container, err := testcontainers.GenericContainer(ctx, genericContainerReq)
+	if container != nil {
+		c.Container = container
+	}
+
+	if err != nil {
+		return c, fmt.Errorf("generic container: %w", err)
+	}
+
+	// only the first node creates the cluster
+	if settings.currentNode == 0 {
+		for i := 1; i < len(settings.nodeNames); i++ {
+			// move to the next node
+			childNode, err := Run(ctx, req.Image, append(clusterOpts, withCurrentNode(i))...)
+			if err != nil {
+				// return the parent cluster node and the error, so the caller can clean up.
+				return c, fmt.Errorf("run cluster node: %w", err)
+			}
+
+			c.childNodes = append(c.childNodes, childNode)
+		}
+	}
+
+	c.opts = settings
+
+	return c, nil
+}
+
+// configureCluster configures the cluster settings, ensuring that the cluster is properly configured with the necessary network and options,
+// avoiding duplicate application of options to be passed to the successive nodes.
+func configureCluster(ctx context.Context, settings *options, opts []testcontainers.ContainerCustomizer) ([]testcontainers.ContainerCustomizer, error) {
+	var clusterOpts []testcontainers.ContainerCustomizer
+	if len(settings.nodeNames) <= 1 {
```

Comment:
suggestion: use zero now we have made the change to WithNodes
```suggestion
	if len(settings.nodeNames) == 0 {
```